### ISSUE TITLE
False Positive: BC_UNCONFIRMED_CAST_OF_RETURN_VALUE when using generic bounded type

### ISSUE THREAD

**Issue by pcimcioch** [spotbugs/spotbugs#1219]
Following code is safe in terms of casting, but spotbugs reports `BC_UNCONFIRMED_CAST_OF_RETURN_VALUE` 
```
package test;

class Parent {}
class Child extends Parent {}

class Factory<T extends Parent> {
    private final Class<T> type;

    Factory(Class<T> type) {
        this.type = type;
    }

    public T create() {
        try {
            return type.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            return null;
        }
    }
}

public class Test {
    public static void main(String[] args) {
        Factory<Child> cs = new Factory<>(Child.class);
        Child child = cs.create(); // BC_UNCONFIRMED_CAST_OF_RETURN_VALUE
        System.out.println(child);
    }
}
```

- If `Parent` is interface instead of class, this bug is **not reported**
```
interface Parent {}
class Child implements Parent {}
```
- If `Factory` doesn't bound type T to `Parent`, this bug is **not reported**
```
class Factory<T> {
```
- If type is bounded on method-level, this bug is **not reported**
```
class Factory {
    public <T extends Parent> T create(Class<T> type) {
        try {
            return type.newInstance();
        } catch (InstantiationException | IllegalAccessException e) {
            return null;
        }
    }
}
```

Spotbugs Version: 4.0.6
Report Level: Low

**Comment by alzimmermsft** [comment]
I'm also seeing a similar issue as this with a method such as:

```java
public <T extends Throwable> T logThrowableAsError(T throwable) {
    // Logic to log the throwable.
    return throwable;
}

// Other code later
public void attemptSomething() throws IOException {
    try {
        trySomething();
    } catch (IOException ex) {
        throw logThrowableAsError(ex);
    }
}
```

**Comment by uhafner** [comment]
No, I am running latest 4.4.2 and the bug still is there:
https://ci.jenkins.io/job/Plugins/job/warnings-ng-plugin/job/PR-1075/15/spotbugs/source.011b3afe-630a-44df-8856-41442c138a42/#177

**Comment by danielnorberg** [comment]
I'm also seeing this issue.

**Comment by zman0900** [comment]
Also seeing this with fairly basic use of Jackson.  For example:

```Java
package crap

import com.fasterxml.jackson.databind.ObjectReader;
import com.fasterxml.jackson.databind.json.JsonMapper;

public final class MyClass {
    private static final ObjectReader JSON_READER = JsonMapper.builder()
            .build()
            .readerFor(OtherClass.class);
}
```

> [ERROR] Low: Unchecked/unconfirmed cast from com.fasterxml.jackson.databind.ObjectMapper to com.fasterxml.jackson.databind.json.JsonMapper of return value in crap.MyClass.<static initializer for MyClass>() [crap.MyClass] At MyClass.java:[line 8] BC_UNCONFIRMED_CAST_OF_RETURN_VALUE

**Comment by patchwork01** [comment]
I've reproduced this here: https://github.com/gchq/sleeper/pull/1162

**Comment by salbracco24** [comment]
Any progress on this?

**Comment by robinjhector** [comment]
`BC_UNCONFIRMED_CAST_OF_RETURN_VALUE` is also apparent when you use the new pattern matching switch in java 21, eg:

```java
SomeInterface myObject = ...

var result = switch(myObject) {
    case Impl1 i1 -> doStuff(i1); //i1 is cast to Impl1
}
```

**Comment by gtoison** [comment]
This should be fixed by #2811 once released

**Comment by JuditKnoll** [comment]
> This should be fixed by #2811 once released

This was released in 4.8.4, the issue should be fixed.
